flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/adaptive/AdaptiveJoinOperatorGenerator.java

```diff
@@ -0,0 +1,184 @@
+                    leftRowCount,
+                    rightRowCount,
+                    tryDistinctBuildRow,
+                    managedMemory,
+                    config,
+                    classLoader);
+        } else {
+            return SorMergeJoinOperatorUtil.generateOperatorFactory(
+                    condFunc,
+                    leftType,
+                    rightType,
+                    leftKeys,
+                    rightKeys,
+                    joinType,
+                    config,
+                    leftIsBuild,
+                    filterNulls,
+                    managedMemory,
+                    classLoader);
+        }
+    }
+
+    @Override
+    public Tuple2<Boolean, Boolean> enrichAndCheckBroadcast(
+            long leftInputSize, long rightInputSize, long threshold) {
+        Tuple2<Boolean, Boolean> isBroadcastAndLeftBuild;
+        boolean leftSizeSmallerThanThreshold = leftInputSize <= threshold;
+        boolean rightSizeSmallerThanThreshold = rightInputSize <= threshold;
+        boolean leftSmallerThanRight = leftInputSize < rightInputSize;
+        switch (joinType) {
+            case RIGHT:
+                // For a right outer join, if the left side can be broadcast, then the left side is
+                // always the build side; otherwise, the smaller side is the build side.
+                isBroadcastAndLeftBuild =
+                        new Tuple2<>(
+                                leftSizeSmallerThanThreshold,
+                                leftSizeSmallerThanThreshold ? true : leftSmallerThanRight);
+                break;
+            case INNER:
+                isBroadcastAndLeftBuild =
+                        new Tuple2<>(
+                                leftSizeSmallerThanThreshold || rightSizeSmallerThanThreshold,
+                                leftSmallerThanRight);
+                break;
+            case LEFT:
+            case SEMI:
+            case ANTI:
+                // For left outer / semi / anti join, if the right side can be broadcast, then the
+                // right side is always the build side; otherwise, the smaller side is the build
+                // side.
+                isBroadcastAndLeftBuild =
+                        new Tuple2<>(
+                                rightSizeSmallerThanThreshold,
+                                rightSizeSmallerThanThreshold ? false : leftSmallerThanRight);
+                break;
+            case FULL:
+            default:
+                throw new RuntimeException(String.format("Unexpected join type %s.", joinType));
+        }
+
+        isBroadcastJoin = isBroadcastAndLeftBuild.f0;
+        leftIsBuild = isBroadcastAndLeftBuild.f1;
+        // Sort merge join requires the left side to be read first if the broadcast threshold is not
+        // met.
+        if (!isBroadcastJoin && originalJoin == OperatorType.SortMergeJoin) {
+            return new Tuple2<>(false, true);
```

Comment:
In the HashJoinOperator, the side that serves as the build side is determined based on leftIsBuild. The SortMergeJoin does not perform this operation, and if a conversion were to be performed, it could lead to a mismatch in input data types.